#129 add community requested source and destination objects

rvveber opened the pull request:
<!--
#########################################################################

Thank you for sharing your work and for opening a PR.

(!) IMPORTANT (!):
First make sure that you point your PR to the `dev` branch!

Now please read the comments carefully and try to provide information
on all relevant titles.

#########################################################################
-->
# Pull Request

## Related Issue

<!--
Add the related issue in the form of #issue-number (Example #100)
-->
closes #129 

## New Behavior

<!--
Please describe in a few words the intentions of your PR.
-->
This Merge Request, enables to specify various new `Source` Entities for the `Standard` and `Extended` ACL Rules.
It additionally enables to specify the same Entities as a `Destination` for the `Extended` ACL Rule.
The new selectable entities are all from the IPAM model group:
- IP-Range
- IP-Address
- Aggregate
- Service

I have built in constraints and validation to prevent assigning more than one source/destination.
Therefore adding community requested entities, while keeping the functionality the same.
No changes to the API, except these newly added fields:
`source_iprange`, `source_ipaddress`, `source_aggregate`, `source_service`
and
`destination_iprange`, `destination_ipaddress`, `destination_aggregate`, `destination_service`
...

## Contrast to Current Behavior

<!--
Please describe in a few words how the new behavior is different
from the current behavior.
-->
Previously you could only assign a Prefix as Source and Destination.
Now you can assign a Prefix, an IP-Range, an IP-Address, an Aggregate or a Service.
...

## Discussion: Benefits and Drawbacks

<!--
Please make your case here:

- Why do you think this project and the community will benefit from your
  proposed change?
  It has been heavily requested by the community and related customers/admins.
- What are the drawbacks of this change?
  There are none, except maybe the addition of repeated code, that could be refactored down in the future.
- Is it backwards-compatible?
  Yes! The original `source_prefix` and `destination_prefix` fields are untouched, they still exist, there are just more 
  `source_*` and `destination_*` fields now.

-->

...

## Changes to the Documentation

<!--
If the docs must be updated, please include the changes in the PR.
If the Wiki must be updated, please make a suggestion below.
-->

...

## Proposed Release Note Entry

<!--
Please provide a short summary of your PR that we can copy & paste
into the release notes.
-->
Added requested models for selection as source/destination: IP-Range, IP-Address, Aggregate, Service.
...

## Double Check

<!--
Please put an x into the brackets (like `[x]`) if you've completed that task.
-->

* [x] I have explained my PR according to the information in the comments
 or in a linked issue.
* [x] My PR targets the `dev` branch.

cruse1977 commented:
Hi @rvveber  - thanks for this, can I ask you take a look at https://netboxlabs.com/docs/netbox/en/stable/plugins/development/migration-v4/ - note the changes to serializers (hence this in conflict currently)

rvveber commented:
> Hi @rvveber - thanks for this, can I ask you take a look at https://netboxlabs.com/docs/netbox/en/stable/plugins/development/migration-v4/ - note the changes to serializers (hence this in conflict currently)

Yes, i will look at it tomorrow!

rvveber commented:
@cruse1977 i rebased onto the new version and updated my commits to solve the conflicts.
After a quick test everything seems to be working as before. You can review

rvveber commented:
@cruse1977 bump

Mailstorm-ctrl commented:
Any update on this? I've bene looking forward to this feature for quite sometime.

cruse1977 commented:
Hi @rvveber just wanted you to know this isn't forgotten about - however its a big change. I've merged this into a forked copy of the latest dev to review - looks good so far.

mrrtfm commented:
> Hi @rvveber just wanted you to know this isn't forgotten about - however its a big change. I've merged this into a forked copy of the latest dev to review - looks good so far.

Yaay, looking forward to this, thanks to all the contributors!

alexis-UE commented:
Hi @rvveber  and @cruse1977  great feature very much needed ! Thanks a lot for the work !!!
Can we merge this one ? Do you need any help on resolving conflicts or something ?

alexis-UE commented:
Hi guys @rvveber @cruse1977, I rebase this PR from the latest version of the `dev` branch and opened this new PR: https://github.com/netbox-community/netbox-acls/pull/236
Are you OK to merge it ?

rvveber commented:
@alexis-UE The rebase isn't the issue.
https://github.com/netbox-community/netbox-acls/issues/129#issuecomment-2591053975 is the issue (if we listen to him).
I don't have the time to re-implement the feature with dynamic assignments.

pheus commented:
Thanks a lot for this great work, @rvveber!

As previously discussed in the issue, we ended up choosing the `assigned_object` approach, so I’m going to close this PR. Even so, your work was a great inspiration, and I really appreciate the time and effort you invested in this. Thanks again for your contribution!
